app-main/public/js/lessons/markdown_export.js

```diff
@@ -0,0 +1,203 @@
+            md += "";
+            break;
+          case "P":
+            md += element.textContent;
+            break;
+          case "IMG":
+            let imageCount = imageCounter();
+            md += `![Image ${imageCount}](${imageCount}.jpg)`;
+            console.log(md);
+            break;
+          case "STRONG":
+            md += `**${element.textContent.trim()}**`;
+            break;
+          case "EM":
+            md += `*${element.textContent.trim()}*`;
+            break;
+          case "PRE":
+            if (element.className === "ql-syntax") {
+              return [
+                { text: "```", newline: true },
+                { text: element.textContent, newline: false },
+                { text: "```", newline: true },
+              ];
+            }
+            break;
+        }
+      }
+      return md;
+    };
+
+    const listToMD = (listParentElement) => {
+      let listItems = [...listParentElement.childNodes];
+      return listItems.map((li, index) => {
+        if (listParentElement.nodeName === "OL") {
+          return `${index + 1}. ${li.textContent} `;
+        } else if (listParentElement.nodeName === "UL") {
+          return `* ${li.textContent} `;
+        }
+      });
+    };
+
+    const headingToMD = (className, textContent) => {
+      switch (className) {
+        case "ql-size-small":
+          return `${textContent} \n`;
+          break;
+        case "ql-size-large":
+          return `<h2>${textContent}</h2>`;
+          break;
+        case "ql-size-huge":
+          return `<h1>${textContent}</h1>`;
+          break;
+      }
+    };
+
+    //used to retrieve children tags out of their <p> tag parents
+    const extractChildren = (element) => {
+      let isList = element.nodeName === "OL" || element.nodeName === "UL";
+      let childElement = hasElementChild(element);
+      if (isList) {
+        return listToMD(element);
+      } else if (typeof childElement === "string") {
+        return mapTagToMD(childElement, markdown);
+      } else {
+        return childElement.length > 0
+          ? mapTagToMD(childElement[0], markdown)
+          : mapTagToMD(element, markdown);
+      }
+    };
+
+    const getImages = (document) => {
+      return [...document.body.querySelectorAll("img")];
+    };
+
+    const extractBase64FromImage = (imageElement) => {
+      return imageElement.src.replace(/data:.+?,/, "");
+    };
+
+    const downloadFile = (filename, data) => {
+      const zip = new JSZip();
+      const images = getImages(parsedDocument);
+      if (images.length > 0) {
+        images.forEach((image, item) => {
+          zip.file(`${item + 1}.jpg`, extractBase64FromImage(image), {
+            base64: true,
+          });
+        });
+        zip.file(
+          `${filename}.md`,
+          data.reduce((string, line) => {
+            return (string += line);
+          }, "")
+        );
+        zip.generateAsync({ type: "blob" }).then(function (content) {
```

Comment:
Do we need to safeguard against this async logic failing here? Could we wrap this in a try/catch or chain a `.catch` block?